gen/api/proxygen.go

```diff
@@ -0,0 +1,297 @@
+package main
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+	"unicode"
+
+	"golang.org/x/xerrors"
+)
+
+type methodMeta struct {
+	node  ast.Node
+	ftype *ast.FuncType
+}
+
+type Visitor struct {
+	Methods map[string]map[string]*methodMeta
+	Include map[string][]string
+}
+
+func (v *Visitor) Visit(node ast.Node) ast.Visitor {
+	st, ok := node.(*ast.TypeSpec)
+	if !ok {
+		return v
+	}
+
+	iface, ok := st.Type.(*ast.InterfaceType)
+	if !ok {
+		return v
+	}
+	if v.Methods[st.Name.Name] == nil {
+		v.Methods[st.Name.Name] = map[string]*methodMeta{}
+	}
+	for _, m := range iface.Methods.List {
+		switch ft := m.Type.(type) {
+		case *ast.Ident:
+			v.Include[st.Name.Name] = append(v.Include[st.Name.Name], ft.Name)
+		case *ast.FuncType:
+			v.Methods[st.Name.Name][m.Names[0].Name] = &methodMeta{
+				node:  m,
+				ftype: ft,
+			}
+		}
+	}
+
+	return v
+}
+func main() {
+	if err := runMain(); err != nil {
+		fmt.Println("error: ", err)
+	}
+}
+
+func typeName(e ast.Expr) (string, error) {
+	switch t := e.(type) {
+	case *ast.SelectorExpr:
+		return t.X.(*ast.Ident).Name + "." + t.Sel.Name, nil
+	case *ast.Ident:
+		pstr := t.Name
+		if !unicode.IsLower(rune(pstr[0])) {
+			pstr = "api." + pstr // todo src pkg name
+		}
+		return pstr, nil
+	case *ast.ArrayType:
+		subt, err := typeName(t.Elt)
+		if err != nil {
+			return "", err
+		}
+		return "[]" + subt, nil
+	case *ast.StarExpr:
+		subt, err := typeName(t.X)
+		if err != nil {
+			return "", err
+		}
+		return "*" + subt, nil
+	case *ast.MapType:
+		k, err := typeName(t.Key)
+		if err != nil {
+			return "", err
+		}
+		v, err := typeName(t.Value)
+		if err != nil {
+			return "", err
+		}
+		return "map[" + k + "]" + v, nil
+	case *ast.StructType:
+		if len(t.Fields.List) != 0 {
+			return "", xerrors.Errorf("can't struct")
+		}
+		return "struct{}", nil
+	case *ast.InterfaceType:
+		if len(t.Methods.List) != 0 {
+			return "", xerrors.Errorf("can't interface")
+		}
+		return "interface{}", nil
+	case *ast.ChanType:
+		subt, err := typeName(t.Value)
+		if err != nil {
+			return "", err
+		}
+		if t.Dir == ast.SEND {
+			subt = "->chan " + subt
+		} else {
+			subt = "<-chan " + subt
+		}
```

Comment:
how do we handle bidirectional channels? Or are they just not a thing in the api?

Reply:
Those don't work over jsonrpc